File: rocket-bi-web/di-core/common/domain/model/user/UserProfile.ts
/*
 * @author: tvc12 - Thien Vi
 * @created: 5/30/21, 10:23 PM
 */

import { Gender, UserGenders } from '@core/common/domain/model';
import { TrackingProfile } from '@core/tracking/domain/TrackingProfile';
import { HtmlElementRenderUtils } from '@/utils/HtmlElementRenderUtils';
import { DateTimeFormatter } from '@/utils';
import { StringUtils } from '@/utils/StringUtils';
import { isObject } from 'lodash';

export class UserProfile {
  username: string;
  alreadyConfirmed: boolean;
  fullName?: string;
  lastName?: string;
  firstName?: string;
  email?: string;
  mobilePhone?: string;
  gender?: number;
  dob?: number;
  avatar?: string;
  oauthType?: string;
  properties?: { [key: string]: string };
  updatedTime?: number;
  createdTime?: number;

  constructor(object: any) {
    this.username = object.username ?? '';
    this.alreadyConfirmed = object.alreadyConfirmed ?? false;
    this.fullName = object.fullName;
    this.lastName = object.lastName;
    this.firstName = object.firstName;
    this.email = object.email;
    this.mobilePhone = object.mobilePhone;
    this.gender = UserGenders.toGenderId(object.gender ?? UserGenders.Other);
    this.dob = object.dob;
    this.avatar = object.avatar;
    this.oauthType = object.oauthType;
    this.properties = object.properties;
    this.updatedTime = object.updatedTime;
    this.createdTime = object.createdTime;
  }

  static unknown() {
    return new UserProfile({
      fullName: 'Unknown'
    });
  }

  get userAvatar(): string {
    return this.avatar && StringUtils.isNotEmpty(this.avatar) ? this.avatar : this.defaultAvatar;
  }

  get defaultAvatar(): string {
    return HtmlElementRenderUtils.renderAvatarAsDataUrl(this.fullName ?? '') ?? '';
  }

  get displayDateOfBirth(): string {
    return this.dob ? DateTimeFormatter.formatAsDDMMYYYY(this.dob) : '';
  }

  displayExtraFieldName(fieldName: string): string {
    return StringUtils.camelToCapitalizedStr(fieldName);
  }

  get displayGender(): string {
    return UserGenders.toDisplayName(this.gender ?? UserGenders.Other);
  }

  static fromObject(object: any): UserProfile {
    return new UserProfile(object);
  }

  get getName(): string {
    return this.fullName || `${this.firstName || ''} ${this.lastName || ''}`.trim() || this.email || this.username;
  }

  static toTrackingProfile(userProfile: UserProfile): TrackingProfile {
    return new TrackingProfile(
      userProfile.username ?? '',
      userProfile.fullName ?? '',
      userProfile.firstName ?? '',
      userProfile.lastName ?? '',
      userProfile.email ?? '',
      userProfile.dob ?? 0,
      userProfile.createdTime ?? 0,
      userProfile.updatedTime ?? 0,
      '',
      '',
      '',
      userProfile.mobilePhone ?? '',
      UserGenders.toDisplayName(userProfile.gender ?? UserGenders.Other),
      userProfile.avatar ?? '',
      userProfile.properties
    );
  }
}

File: server/src/main/scala/co/datainsider/caas/user_profile/domain/user/UserProfile.scala
package co.datainsider.caas.user_profile.domain.user

/**
  * @author anhlt
  */
object UserGender {
  val Other = -1
  val Female = 0
  val Male = 1
}

case class ShortUserProfile(
    username: String,
    fullName: Option[String],
    lastName: Option[String],
    firstName: Option[String],
    gender: Option[Int],
    avatar: Option[String]
)

@SerialVersionUID(1L)
case class UserProfile(
    username: String,
    fullName: Option[String] = None,
    lastName: Option[String] = None,
    firstName: Option[String] = None,
    email: Option[String] = None,
    mobilePhone: Option[String] = None,
    gender: Option[Int] = None,
    dob: Option[Long] = None,
    avatar: Option[String] = None,
    alreadyConfirmed: Boolean = false,
    properties: Option[Map[String, String]] = None,
    updatedTime: Option[Long] = None,
    createdTime: Option[Long] = None
) {
  def toShortUserProfile: ShortUserProfile = {
    ShortUserProfile(
      username = username,
      fullName = fullName,
      lastName = lastName,
      firstName = firstName,
      gender = gender,
      avatar = avatar
    )
  }
}
